### ISSUE TITLE
[MOBILE] Mobile Support - Global Tracking Issue

### ISSUE THREAD

**Issue by guillemcordoba** [holochain/holochain#3243]
After a full iteration of development + deployment + testing cycle with [the rostanga mobile app](https://github.com/darksoil-studio/rostanga), I've explored a lot of the territory for running holochain on mobile. Right now I'm actually having to maintain a [fork of holochain that stores all data in firebase](https://github.com/guillemcordoba/holochain/blob/ff3fe54d819ae9cf791c5c3798ccb0c9217884f6/crates/holochain_p2p/src/spawn/actor.rs#L371) because of the tx5 `PendingJoin` bug that is a known issue in 0.3.0. Also this makes me less nervous about holochain processes running into [these kinds of issues](https://github.com/holochain/holochain/issues/2742) while running for full days.

Anyway, in my view this is a comprehensive list of the issues that holochain core + its dependencies need to fix before we can say that "holochain supports mobile":

## UX issues

- [ ] Holochain takes a long time to boot up _every time you open the app_ (10 seconds on my android device):
  - [ ] 5 seconds of that is running the lair keystore server: https://github.com/holochain/lair/issues/120
- [ ] It takes a long time to install apps (in the order of 40 seconds)
  - [ ] Make sure that the install process is as streamlined as possible
    - [ ]  https://github.com/holochain/holochain/pull/3242

## Networking

- [ ] Tx5 `PendingJoin` known issue
- [ ] https://github.com/holochain/tx5/issues/79

## Push Notifications

To achieve push notifications in mobile OSes, we basically need to go through Firebase Cloud Messaging (FCM) and Apple Notification Service (APN). Here I'm assuming tauri as the mobile runtime that holochain is running in.

- [ ] I've forked the [notification plugin for tauri](https://github.com/guillemcordoba/plugins-workspace/tree/v2/plugins/notification) 
  - [ ] I've forked [tao](https://github.com/guillemcordoba/tao) to be able to set up the notification plugin. That needs to be merged into the main repository. 
  - [ ] I've forked [swift-rs](https://github.com/guillemcordoba/swift-rs) to fix a compilation issue. That needs to be merged into the main repository.
- [x] I've created a PoC of the semi-centralized bridge service to be able to actually send the notifications to offline agents: https://github.com/darksoil-studio/notifications

## Android specific issues

- [ ] I'm seeing [this issue](https://github.com/wasmerio/wasmer/issues/2861) intermitently show up after a minute or so of using the app and calling the conductor. More investigation is needed.

### Precompiled wasms

This is the area that I have explored the least. So I'm not sure this is ready to go as it is in the code. I do now that @Connoropolous played a big role in the implementation of prepackaged wasms in iOS, so I'll let him comment if he sees fit to do so. But at least we need a:

- [ ] Good guide on how to precompile and package your wasms in dylibs / iOS frameworks and package it in your iOS app.

### Interpreted wasms

AFAIK, [iOS has as a rule that you can't execute or JIT compile anything that wasn't prepackaged as part of your iOS app at compile time](https://github.com/wasmerio/wasmer/issues/4343).

> Note: be careful, as apple actually lets you do things when developing with xcode that then you can't do when publishing an app to the appstore. In my development, holochain worked fine by JIT compiling the wasms, but that stopped working only when I build the "production" app build and downloaded it to the appstore.

`wasmer` supports a bunch of JIT compile engines, one of them is JavascriptCore which uses the underlying OS based JIT, **but** in which [apple have dropped support for webassembly and will not be bringing it back](https://bugs.webkit.org/show_bug.cgi?id=265151#c17). I've done some research and `wasmer` had [a vote](https://github.com/wasmerio/wasmer/discussions/3839) around implementing more backends, one of which is wasmi, which is a wasm interpreter. They could also implement something like a bridge to `WKWebView` which does seem to have native `WebAssembly` support in iOS, but I haven't seem any mention of that in any of their repositories or docs.

With all of this, **we are left with no path to implement a runtime like `we`, `neighborhoods` or the `launcher` in iOS**. These are runtimes that don't prepackage all installable apps in them, but rather they download them from the internet and install them dynamically in the conductor. I certainly want to enable these kinds of use cases, but it seems like without a big effort they are not possible with the current state of things. 

The solutions that I can see are:

- We could re-add wasmi as a possible wasm engine behind a feature flag
- We could help wasmer implement an interpreter backend
- We could help wasmer implement a bridge to `WKWebView` 

All of these seem like a big lift, but this seems crucially important. @thedavidmeister am I missing something in this territory?

I'd be really interested in talking about more options in this territory.

---

I'll try to keep this issue up to date with all the fixes and developments so that we have an overall view of the state of affairs.

**Comment by Connoropolous** [comment]
hiyas...
this line explains a bit about how the precompiled thing worked, but I have no idea anymore about how wasmer has moved. 
https://github.com/holochain/holochain/blob/194bc259a4d683d83e4953bb24bf5e6e8b7294fe/CHANGELOG.md?plain=1#L2116

There was the additional step that during the Dna read phase you needed to translate the `dylib` "paths" and read in the bytes. That's in a private repo.

**Comment by guillemcordoba** [comment]
@Connoropolous thanks! did you need to package your dylib deserialized wasms into iOS frameworks for it to work?

**Comment by Connoropolous** [comment]
I needed to have those dylib files sitting in the Xcode project folder and make sure they were included on the bundle of things that got codesigned during the app build process. I also needed to develop by testing and running on a hardware device, no simulator.

**Comment by guillemcordoba** [comment]
That's really helpful @Connoropolous , thanks!

**Comment by abe-njama** [comment]
@guillemcordoba what's this Tx5 `PendingJoin` known issue? Could you be referring to the performance issue on Android 33?

**Comment by guillemcordoba** [comment]
Oh sorry, it's not really tx5 as I got yesterday, it's the issue that @maackle is referring to [here](https://github.com/holochain/holochain/discussions/3252), that makes 0.3.x not viable right now, because when you enable the app is app is not enabled and stays in paused mode with `PendingJoin` as its reason.

**Comment by maackle** [comment]
FYI https://github.com/holochain/holochain/pull/3221 just merged, removing the network join error

**Comment by thedavidmeister** [comment]
@guillemcordoba is adding wasmi as an engine really that big of a lift?

i'd expect wasmer to basically work as-is because we'd still be using its interface, with wasmi only implemented "under the hood" of wasmer, right?

**Comment by guillemcordoba** [comment]
@thedavidmeister I don't think it should be that much of a lift no, and for holochain I think that would be the ideal, since then from holochain's point of view it's still wasmer.

Problem is I haven't seen any movement whatsoever in any wasmer repos/discord/discussions/pullrequests/articles that indicate to me that adding wasmi as an engine is in their backlog. So we'd have to request for that I imagine, and put pressure on it so that it is implemented... And meanwhile we have a bunch of time constraints on our side of needing to build products that work now on iOS (at least this is the situation that I'm in right now).

**Comment by thedavidmeister** [comment]
@guillemcordoba hmm, i have no idea what is involved in making an engine or who to talk to about it, do you?

**Comment by guillemcordoba** [comment]
Nope... No idea, the wasmer repo and discord is the first thing I would try.

**Comment by abe-njama** [comment]
@zippy @guillemcordoba @thedavidmeister

I've filed an [issue](https://github.com/wasmerio/wasmer/issues/4432) in the wasmer repo requesting support or a support model be proposed by the wasmer team so we can confidently work on any one of the Interpreted wasm solutions being suggested by @guillemcordoba in this main issue. This is a follow-up to an email that I'd sent to the wasmer maintainers but I hadn't received a response to. 
Please feel free to pile on to the issue so we can get some traction.

**Comment by abe-njama** [comment]
Hey @guillemcordoba. Are you able to update the test application to 'holochain 0.3.0-beta-dev.37' and run the tests again? The fixes to some of the UX and Networking issues that had been raised have been shipped and we'd like to understand what impact these have had and to take stock of any new issues and to keep focus on outstanding ones.
- [Relax our notion of "joining the network"](https://github.com/holochain/holochain/pull/3221)
- [5s network join timeout](https://github.com/holochain/holochain/pull/3263)
- [Move wasmer module cache to conductor](https://github.com/holochain/holochain/pull/3242)
- [Refactor of endpoint and connection management state logic](https://github.com/holochain/tx5/pull/77) and its integration to Holochain; [Tx5 integration of endpoint/connection state management refactor](https://github.com/holochain/holochain/pull/3287)

**Comment by mattyg** [comment]
Hey @guillemcordoba -- I'm trying to clean this up to make sure everything has a standalone issue in the #4340 Epic

Any progress on getting your changes to tao and tauri-plugin-notification merged upstream? 

Do you think this is resolved? I don't think I've seen it: https://github.com/wasmerio/wasmer/issues/2861
